### ISSUE TITLE
Interceptors downstream of async interceptor are executed on the core.async thread pool

### ISSUE THREAD

**Issue by mullr** [pedestal/pedestal#550]
# Description
If you put an async interceptor into your interceptor chain, subsequent interceptors are executed on the core.async thread pool. Clearly they have to happen /somewhere/, but this particular behaviour is very surprising and may lead to uncomfortable unintended consequences if something downstream is doing I/O. 

This is at least a documentation bug; I'd like to see this spelled out in bright flashing letters somewhere. 

As a fix, it might be nice be to allow (require?) specifying a thread pool to be used for subsequent requests. 

## Steps to reproduce
Use an interceptor chain like this:
```clojure
{:name ::upstream
 :enter (fn [context]
          (println "---------- UPSTREAM ENTER ------------")
          (log/error (ex-info "" {}) "")
          context)}
{:name ::async-thing
 :enter (fn [context]
          (async/go
            (println "************************")
            context))}
{:name ::downstream
 :enter (fn [context]
          (println "---------- DOWNSTREAM ENTER ------------")
          (log/error (ex-info "" {}) "")
          context)}
```

## Pedestal version
0.5.2

**Comment by ddeaguiar** [comment]
@ohpauleez, `go` blocks are used to resume execution once the async interceptor returns a context. Refer to the [`go-async`](https://github.com/pedestal/pedestal/blob/master/interceptor/src/io/pedestal/interceptor/chain.clj#L110-L144) implementation. I don't think they can be removed without replacing them with something else.

**Comment by hlship** [comment]
I'm very interested in seeing this fixed as we're starting to introduce asynchronous interceptors into our pipeline.  We've had to up the value for `clojure.core.async.pool-size` from its default to allow for blocking code executing inside the dispatch thread pool.

**Comment by timewald** [comment]
It isn't clear to me that moving the work back onto the servlet's async
thread pool solves the issue if the goal is to execute long running
blocking code.

It could be made to dispatch onto a provided executor thread pool, but
there are challenges there too wrt not blocking those threads vs using a
pool that grows arbitrarily large, can you allocate across multiple
different pools for different work streams, etc.

The theory when we first implemented go-async was that having "gone async",
your application logic would do whatever long running work you had to do
using whatever combination of threading and/or async handling you required,
then return the result to finish the interceptor chain processing. The
assumption was that the core.async pool would only be used for
fast-executing HTTP response processing.

That idea still holds for interceptors that are not terminal handlers. So
one approach you could take if you have an interceptor that needs to do
blocking work is to make it an async interceptor itself, in case it is used
in conjunction with another async interceptor that has caused work to move
onto the core.async thread pool. It would do it's work in some task
appropriate way using whatever concurrency constructs it wants. When the
work completes it would resume interceptor chain execution.

I totally agree with Russell that the current behavior should be documented.

**Comment by hlship** [comment]
Maybe I interpreted the description wrong; what I got from it was that when an interceptor goes async (returns a channel), all future work for that interceptor chain occurs in a core.async dispatch thread (of which there are, by default, only 8).

In our particular application, we generally see request processing time maxing out around 2000 ms, mostly spent in that terminal handler (which is not coded as async).  However, if a dispatch thread is occupied executing that code, then that means that (potentially) other CSPs that should be executing on a dispatch thread are blocked or delayed.

I have not dived down into the code, so this may be all smoke, no fire.

**Comment by timewald** [comment]
Your understanding is correct, that is exactly what happens.

This is no different than if you combine multiple core.async-based CSPs
that use the async thread pool. You have to tune the async thread pool size
to work with the combination of libraries you've chosen. CSP code that uses
the async thread pool should not be doing long running blocking work on
those threads.

In your case, if the bulk of the blocking work is in the handler, you would
only be blocking the async thread pool if an earlier interceptor already
went async and switched execution over to that pool. If you make your
handler async, you can dispatch the blocking work onto whatever thread pool
you like without interfering with anything else.

If pedestal mapped resumed async execution onto the servlet async pool as
Paul suggested, it would present the same issue - you don't want to block
those threads for extended periods either.

Pedestal could be modified to use a thread pool of its own, or a supplied
executor service, but that presents challenges in its own rights - starting
with not necessarily wanting to have another thread pool warm in your
process.

From my perspective, I think pedestal already provides the infrastructure
that's required to handle the situation you describe, but we do need to fix
the documentation.

**Comment by ohpauleez** [comment]
I'll also add that Pedestal's infrastructure and decisions are very similar to those made in Java's CompleteableFuture.

I think the only fix here is documentation.

**Comment by SherSpock** [comment]
Is this waiting for somebody? Seems like an easy first issue for a newcomer such as myself.

**Comment by ohpauleez** [comment]
I always love to see people get involved!  Feel free to submit documentation updates as you like.  The docs live in their own repository -- https://github.com/pedestal/pedestal-docs

**Comment by ddeaguiar** [comment]
@SherSpock, [here's](https://github.com/pedestal/pedestal-docs/issues/104) the related pedestal-docs issue!

**Comment by SherSpock** [comment]
@ohpauleez @ddeaguiar Thanks much! Looking

**Comment by SherSpock** [comment]
Opened pedestal/pedestal-docs/pull/126

Anything missing or undesirable?

**Comment by SherSpock** [comment]
> There are already [interceptor chain tests](https://github.com/pedestal/pedestal/blob/master/service/test/io/pedestal/interceptor_test.clj) that capture the correct flow, but **we should add a few tests that track thread IDs to better document the locality of execution.**

@ohpauleez Is this an open issue already?

**Comment by SherSpock** [comment]
@ddeaguiar I opened pedestal/pedestal-docs/pull/126 to address this issue (#550) rather than addressing [#104](https://github.com/pedestal/pedestal-docs/issues/104) on the docs side. Though clearly related and arguably a clear superset, #104 is dramatically larger in scope. 

What I think this needs is a new issue on the docs side for just the warning, which I have opened: pedestal/pedestal-docs/issues/128. 

I would respectfully suggest this issue remain open until such time as a warning is added on this side of the house or such a warning is deemed unnecessary.
